### PR TITLE
Improve the way how the operator handles partitial upgrades

### PR DESCRIPTION
**Pull request by johscheuer**
# Description

This change makes sure we are not setting any FDB environment variables for the `fdbcli` command to prevent the use of the multi version client for 7.1+ fdbclis. In addition we added more fallback logic to the `GetStatus` method to transparently try the new (desired) version if the runningVersion is not working.

## Type of change

*Please select one of the options below.*

- Bug fix (non-breaking change which fixes an issue)

## Discussion

This adds some more complexity around the upgrade handling but the multi version client is not reliable in cases where not all coordinators are upgraded.

## Testing

I will see if I can add some additional unit tests and I will do a test with out new upgrade test.

## Documentation

Will be covered in the other upgrade doc issue.

## Follow-up

-

### CODE REVIEW

**Review by johscheuer — COMMENTED** [review]
I found some other areas that needs some improvement for all the edge cases.

**Comment by johscheuer** [comment]
Ran the test suite with the new test that keeps one coordinator at the old version:

File: fdbclient/admin_client.go [hunk 1]
```diff
@@ -318,7 +309,30 @@ func (client *cliAdminClient) getStatus(useClientLibrary bool) (*fdbv1beta2.Foun
...
+// the new version for fdbcli will be tried.
+func (client *cliAdminClient) getStatus() (*fdbv1beta2.FoundationDBStatus, error) {
+	status, err := client.getStatusFromCli()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the status doesn't contain any processes and we are doing an upgrade, we probably use the wrong fdbcli version
+	// and we have to fallback to the on specified in out spec.version.
+	if len(status.Cluster.Processes) == 0 && client.Cluster.IsBeingUpgraded() {
```
Comment (brownleej) [inline fdbclient/admin_client.go:328]: It looks like this has less logging in this scenario than the existing code, which worries me because of how that logging has been helpful in recent debugging. Can we add that logging back? We might even want to expand it to include the invalid status we got from the client library to give more context.

File: fdbclient/command_runner.go [hunk 2]
```diff
@@ -37,28 +40,61 @@ type realCommandRunner struct {
 	log logr.Logger
 }
 
+// getEnvironmentVariablesWithoutBlacklisted returns the current environment variables for the new process with some
```
Comment (brownleej) [inline fdbclient/command_runner.go:43]:
```suggestion
// getEnvironmentVariablesWithoutExcludedFdbEnv returns the current environment variables for the new process with some
```

File: fdbclient/command_runner.go [hunk 3]
```diff
@@ -37,28 +40,61 @@ type realCommandRunner struct {
...
+// FDB specific variables filtered out to ensure we don't set any variables that could change the behaviour of fdbcli or
+// the other fdb tools.
+func getEnvironmentVariablesWithoutExcludedFdbEnv() []string {
+	blackListedEnvironmentVariables := map[string]fdbv1beta2.None{
```
Comment (brownleej) [inline fdbclient/command_runner.go:47]: I think we generally try to avoid the term "black list" for these kind of use cases. Can we go with something like "blocked" or "excluded" instead?

**Review by brownleej — CHANGES_REQUESTED** [review]

File: fdbclient/admin_client.go (hunk 1 above)
Comment (johscheuer) [reply fdbclient/admin_client.go:328]: I don't think that this scenario has less logging than before. We still report the `"Fetched status JSON with fdbcli"` log line including the used version. In addition we still log the command that is executed including the full binary path and the timeout parameter and I haven't removed any log statements. What logging statement exactly are you missing/looking for?

File: fdbclient/command_runner.go (hunk 3 above)
Comment (johscheuer) [reply fdbclient/command_runner.go:47]: Yes, I adjusted the method name but forgot about the variable name 🤦

File: fdbclient/admin_client.go (hunk 1 above)
Comment (johscheuer) [reply fdbclient/admin_client.go:328]: I added some additional logging.

File: fdbclient/admin_client.go [hunk 4]
```diff
@@ -238,9 +238,10 @@ func (client *cliAdminClient) runCommand(command cliCommand) (string, error) {
 
 		// If we hit a timeout report it as a timeout error
 		if strings.Contains(string(output), "Specified timeout reached") {
+			client.log.Info("runner has timeout issue")
```
Comment (johscheuer) [inline fdbclient/admin_client.go:241]: We can drop this log statement again, I only used it for debugging.

File: fdbclient/admin_client.go [hunk 5]
```diff
@@ -318,7 +309,30 @@ func (client *cliAdminClient) getStatus(useClientLibrary bool) (*fdbv1beta2.Foun
 	if err != nil {
 		return nil, err
 	}
-	client.log.V(1).Info("Fetched status JSON", "status", status)
+
+	client.log.V(1).Info("Fetched status JSON with fdbcli", "version", client.Cluster.GetRunningVersion(), "status", status)
```
Comment (johscheuer) [inline fdbclient/admin_client.go:313]:
```suggestion
	client.log.V(1).Info("Fetched status JSON with fdbcli", "version", client.Cluster.GetRunningVersion())
```

We already log the full status at the return of the GetStatus method.

File: fdbclient/admin_client.go (hunk 1 above)
Comment (johscheuer) [inline fdbclient/admin_client.go:328]:
```suggestion
	if (status == nil || len(status.Cluster.Processes) == 0) && client.Cluster.IsBeingUpgraded() {
```

File: fdbclient/common.go [hunk 6]
```diff
@@ -142,8 +143,19 @@ func getConnectionStringFromDB(cluster *fdbv1beta2.FoundationDBCluster, log logr
...
+		return nil, err
+	}
+
+	status := &fdbv1beta2.FoundationDBStatus{}
+	err = json.Unmarshal(contents, status)
+	if err != nil {
+		return nil, err
+	}
+
+	return status, err
```
Comment (johscheuer) [inline fdbclient/common.go:158]:
```suggestion
	return status, nil
```

**Review by brownleej — APPROVED** [review]
👍

**Comment by sbodagala** [comment]
> This change makes sure we are not setting any FDB environment variables for the fdbcli command to prevent the use of the multi version client for 7.1+ fdbclis.

Why are we doing this?

File: fdbclient/admin_client.go [hunk 7]
```diff
@@ -318,7 +308,30 @@ func (client *cliAdminClient) getStatus(useClientLibrary bool) (*fdbv1beta2.Foun
...
+
+	return status, nil
+}
+
+// getStatus uses fdbcli to connect to the FDB cluster, if the cluster is upgraded and the initial version returns no processes
```
Comment (sbodagala) [inline fdbclient/admin_client.go:315]: You don't want to handle the partial upgrade scenario (where a subset of processes are still/stuck at the older version for whatever reason) here?

It appears you are trying to implement (a version of) the multiversion client logic here, probably because of a limitation of that logic? Maybe you can explain that limitation and how you are addressing that here (so we can follow what you are trying to do)? Thanks!

Comment (johscheuer) [reply fdbclient/admin_client.go:315]: > You don't want to handle the partial upgrade scenario (where a subset of processes are still/stuck at the older version for whatever reason) here?

What scenario do you have in mind? As far as I understand the logic in FDB the majority of coordinators will define the version of the cluster e.g. a cluster can only be at one version (and have a few other processes running at a different version but not reporting to the cluster). We observed a few cases where the kill was not restarting one coordinator (for different reasons) and that resulted in the multi version client behaving badly.

> It appears you are trying to implement (a version of) the multiversion client logic here, probably because of a limitation of that logic? Maybe you can explain that limitation and how you are addressing that here (so we can follow what you are trying to do)? Thanks!

We are implementing it because the multi version logic is currently broken (I'm not sure if that is limited to 6.3 and is fixed with 7.1), there are some comments about this in the FDB Slack. Basically by using the fdbcli binary without multi version configuration we force the operator to use a specific version to connect to the DB.

**Review by johscheuer — COMMENTED** [review]
>> This change makes sure we are not setting any FDB environment variables for the fdbcli command to prevent the use of the multi version client for 7.1+ fdbclis.

> Why are we doing this?

Otherwise `fdbcli` will use the multi-version client with the same result as if we use the multi version client from the operator (go bindings). I think the multi version support for fdbcli was added in 7.1.

File: fdbclient/admin_client.go [hunk 8]
```diff
@@ -291,24 +291,14 @@ func (client *cliAdminClient) runCommandWithBackoff(command string) (string, err
 	return rawResult, err
 }
 
-func (client *cliAdminClient) getStatus(useClientLibrary bool) (*fdbv1beta2.FoundationDBStatus, error) {
-	var contents []byte
-	var err error
-
-	if useClientLibrary {
```
Comment (sbodagala) [inline fdbclient/admin_client.go:298]: So "useClientLibrary" is not used in the modified version, right?

File: fdbclient/command_runner.go [hunk 9]
```diff
@@ -37,9 +40,32 @@ type realCommandRunner struct {
 	log logr.Logger
 }
 
+// getEnvironmentVariablesWithoutBlacklisted returns the current environment variables for the new process with some
+// FDB specific variables filtered out to ensure we don't set any variables that could change the behaviour of fdbcli or
+// the other fdb tools.
+func getEnvironmentVariablesWithoutExcludedFdbEnv() []string {
```
Comment (sbodagala) [inline fdbclient/command_runner.go:46]: Do you want to exclude these variables only when running "status" command? You are doing it for all commands below.
